effectful/handlers/llm/sampling.py

```diff
@@ -45,3 +47,55 @@ def n_votes_ahead():
                 tasks.append(executor.submit(interpreter(intp)(fwd), *args, **kwargs))
         executor.shutdown()
         return self.votes.most_common(1)[0][0]
+
+
+def sample(template, n):
+    @defop
+    def in_nested_call() -> bool:
+        return False
+
+    lock = threading.Lock()
+
+    def _template_call(template, *args, **kwargs):
+        if in_nested_call():
+            return fwd()
+
+        with handler({in_nested_call: lambda: True}):
+            with ThreadPoolExecutor() as executor:
+
+                @interpreter(get_interpretation())
+                def do_work():
+                    lock.acquire()
+                    try:
+                        result = fwd()
+                    finally:
+                        assert lock.locked()
+                        lock.release()
+                    return result
+
+                tasks = [executor.submit(do_work) for _ in range(n)]
+                completed = futures.wait(tasks, return_when=futures.ALL_COMPLETED)
+                return [t.result() for t in completed.done]
+
+    def _completion(*args, **kwargs):
+        lock.release()
+        result = fwd()
+        lock.acquire()
+        return result
+
+    def _tool_call(*args, **kwargs):
+        lock.acquire()
```

Comment:
what exactly are the guarantees that the `lock` here is intended to provide? is it that end users can think of `sample` as being equivalent to running template `n` times in sequence (in a non-deterministic order)?